File: Valid Parentheses/Leetcode_20.py
# top 2%
class Solution:
    def isValid(self, s: str) -> bool:
        d = {')':'(','}':'{',']':'['}
        stack = []
        i = 0
        while i < len(s):
            if s[i] in d.values():
                stack.append(s[i])
            elif stack:
                if stack[-1] != d[s[i]]:
                    return False
                else:
                    del stack[-1]
            else:
                return False
            i+=1
        return not stack                

File: Valid Parentheses/Leetcode_20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> parens;
        for (const auto& c: s) {
            switch(c) {
                case '(': parens.push(')'); break;
                case '[': parens.push(']'); break;
                case '{': parens.push('}'); break;
                default:
                    if (parens.empty() || c != parens.top()) return false;
                    else parens.pop();
            }
        }
        return parens.empty();
    }
};
